src/collections/maple_tree/mod.rs

```diff
@@ -104,66 +131,307 @@ impl<K, V> SparseLeaf<K, V>
             }
         }
 
-        // If there is no slot, return a failure?
+        return None;
+    }
+
+    pub fn search(&mut self, k: &K) -> Option<&V> {
+        for i in 0..CAPACITY {
+            match &self.key[i] {
+                M::Some(v) => {
+                    if v == k {
+                        match &self.value[i] {
+                            M::Some(v) => {
+                                return Some(v);
+                            }
+                            M::None => {
+                                return None;
+                            }
+                        }
+                    }
+                }
+
+                M::None => {}
+            }
+        }
 
-        unimplemented!();
+        None
     }
 
     pub fn update(&mut self, k: K, v: V) -> Option<V> {
-        unimplemented!();
+        for i in 0..CAPACITY {
+            match &self.key[i] {
+                M::Some(v) => {
+                    if v != &k {
+                        continue;
+                    }
+                }
+
+                M::None => {
+                    continue;
+                }
+            }
+
+            let mut nv = M::Some(v);
+
+            mem::swap(&mut self.value[i], &mut nv);
+
+            match nv {
+                M::Some(v) => return Some(v),
+                M::None => return None,
+            }
+        }
+
+        None
     }
 
     pub fn remove(&mut self, k: &K) -> Option<V> {
         for i in 0..CAPACITY {
             println!("remove: {:?}", self.key[i]);
 
-            if self.key[i] == k {
-                let mut nk = M::None;
-                let mut nv = M::None;
+            match &self.key[i] {
+                M::Some(v) => {
+                    if v != k {
+                        continue;
+                    }
+                }
+                M::None => {
+                    continue;
+                }
+            }
 
-                mem::swap(&mut self.key[i], &mut nk);
-                mem::swap(&mut self.value[i], &mut nv);
-                return Some(nv);
+            let mut nk = M::None;
+            let mut nv = M::None;
+
+            mem::swap(&mut self.key[i], &mut nk);
+            mem::swap(&mut self.value[i], &mut nv);
+
+            match nv {
+                M::Some(v) => {
+                    return Some(v);
+                }
+                M::None => {}
+            }
+        }
+        None
+    }
+   
+    // should this return the first key found that is out of the range?
+    // if so it seems we need K to have the clone trait
+    pub fn check_min_max(&mut self, min: &K, max: &K) -> Option<()>{
```

Comment:
What's the intent of this function? What is it trying to achieve? Is it checking that the min/max is less/greater than the content of this node?